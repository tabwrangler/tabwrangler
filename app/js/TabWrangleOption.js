/* @flow */

import React from 'react';

interface Props {
  onChange: (event: SyntheticInputEvent<HTMLInputElement>) => void;
  selectedOption: string;
}

export default function TabWrangleOption(props: Props) {
  // Declare this dynamically so it is available inside tests. It's not simple to modify globals,
  // like `chrome`, using Jest. Is there a better way to do this? Probably.
  const OPTIONS = [
    { name: 'withDupes', text: chrome.i18n.getMessage('options_dedupe_option_withDupes') },
    {
      name: 'exactURLMatch',
      text: chrome.i18n.getMessage('options_dedupe_option_exactURLMatch'),
    },
    {
      name: 'hostnameAndTitleMatch',
      text: chrome.i18n.getMessage('options_dedupe_option_hostnameAndTitleMatch'),
    },
  ];

  return (
    <>
      <label htmlFor="wrangleOption">
        <strong>{chrome.i18n.getMessage('options_dedupe_label')}</strong>
      </label>
      {OPTIONS.map(option => (
        <div className="form-check" key={option.name}>
          <input
            checked={props.selectedOption === option.name}
            className="form-check-input"
            id={option.name}
            name="wrangleOption"
            onChange={props.onChange}
            type="radio"
            value={option.name}
          />
          <label className="form-check-label" htmlFor={option.name}>
            {option.text}
          </label>
        </div>
      ))}
      <div className="row">
        <div className="col-8 form-text text-muted" style={{ marginBottom: 0 }}>
          <dl style={{ marginBottom: 0 }}>
            <dt>{chrome.i18n.getMessage('options_dedupe_option_withDupes_label')}</dt>
            <dd>{chrome.i18n.getMessage('options_dedupe_option_withDupes_description')}</dd>
            <dt style={{ marginTop: '10px' }}>
              {chrome.i18n.getMessage('options_dedupe_option_exactURLMatch_label')}
            </dt>
            <dd>{chrome.i18n.getMessage('options_dedupe_option_exactURLMatch_description')}</dd>
            <dt style={{ marginTop: '10px' }}>
              {chrome.i18n.getMessage('options_dedupe_option_hostnameAndTitleMatch_label')}
            </dt>
            <dd>
              {chrome.i18n.getMessage('options_dedupe_option_hostnameAndTitleMatch_description')}
            </dd>
          </dl>
        </div>
      </div>
    </>
  );
}
