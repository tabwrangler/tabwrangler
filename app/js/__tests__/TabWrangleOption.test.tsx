import React from "react";
import ReactTestUtils from "react-dom/test-utils";
import TabWrangleOption from "../TabWrangleOption";
import renderer from "react-test-renderer";

describe("TabWrangleOption", () => {
  let ogGetMessage: typeof chrome.i18n.getMessage;

  beforeEach(() => {
    ogGetMessage = chrome.i18n.getMessage;
    chrome.i18n.getMessage = () => "";
  });

  afterEach(() => {
    chrome.i18n.getMessage = ogGetMessage;
  });

  test("renders options with withDupes selected", () => {
    const mockCallback = jest.fn();

    const two = renderer.create(
      <TabWrangleOption onChange={mockCallback} selectedOption="withDupes" />
    );

    expect(two).toMatchSnapshot();
  });

  test("renders options with exactURLMatch selected", () => {
    const mockCallback = jest.fn();

    const two = renderer.create(
      <TabWrangleOption onChange={mockCallback} selectedOption="exactURLMatch" />
    );

    expect(two).toMatchSnapshot();
  });

  test("renders options with hostnameAndTitleMatch selected", () => {
    const mockCallback = jest.fn();

    const two = renderer.create(
      <TabWrangleOption onChange={mockCallback} selectedOption="hostnameAndTitleMatch" />
    );

    expect(two).toMatchSnapshot();
  });

  test("calls onChange handler callback when clicked", () => {
    const mockCallback = jest.fn();

    // Must wrap `TabWrangleOption` in a Composite Component in order to find it using ReactTestUtils.
    class Wrapper extends React.Component {
      render() {
        return (
          <div>
            <TabWrangleOption onChange={mockCallback} selectedOption="hostnameAndTitleMatch" />
          </div>
        );
      }
    }

    const two = ReactTestUtils.renderIntoDocument(<Wrapper />);

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore:next-line
    const buttonNode = ReactTestUtils.scryRenderedDOMComponentsWithTag(two, "input");

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore:next-line
    ReactTestUtils.Simulate.change(buttonNode[1], { target: { checked: true } });

    expect(mockCallback.mock.calls.length).toBe(1);
  });
});
