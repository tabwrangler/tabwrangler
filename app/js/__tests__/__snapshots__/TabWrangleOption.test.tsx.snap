// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabWrangleOption renders options with exactURLMatch selected 1`] = `
[
  <strong>
    Translated&lt;options_dedupe_label&gt;
  </strong>,
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="withDupes"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="withDupes"
    />
    <label
      className="form-check-label"
      htmlFor="withDupes"
    >
      Translated&lt;options_dedupe_option_withDupes&gt;
    </label>
  </div>,
  <div
    className="form-check"
  >
    <input
      checked={true}
      className="form-check-input"
      id="exactURLMatch"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="exactURLMatch"
    />
    <label
      className="form-check-label"
      htmlFor="exactURLMatch"
    >
      Translated&lt;options_dedupe_option_exactURLMatch&gt;
    </label>
  </div>,
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="hostnameAndTitleMatch"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="hostnameAndTitleMatch"
    />
    <label
      className="form-check-label"
      htmlFor="hostnameAndTitleMatch"
    >
      Translated&lt;options_dedupe_option_hostnameAndTitleMatch&gt;
    </label>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-8 form-text text-muted"
      style={
        {
          "marginBottom": 0,
        }
      }
    >
      <dl
        style={
          {
            "marginBottom": 0,
          }
        }
      >
        <dt>
          Translated&lt;options_dedupe_option_withDupes_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_withDupes_description&gt;
        </dd>
        <dt
          style={
            {
              "marginTop": "10px",
            }
          }
        >
          Translated&lt;options_dedupe_option_exactURLMatch_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_exactURLMatch_description&gt;
        </dd>
        <dt
          style={
            {
              "marginTop": "10px",
            }
          }
        >
          Translated&lt;options_dedupe_option_hostnameAndTitleMatch_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_hostnameAndTitleMatch_description&gt;
        </dd>
      </dl>
    </div>
  </div>,
]
`;

exports[`TabWrangleOption renders options with hostnameAndTitleMatch selected 1`] = `
[
  <strong>
    Translated&lt;options_dedupe_label&gt;
  </strong>,
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="withDupes"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="withDupes"
    />
    <label
      className="form-check-label"
      htmlFor="withDupes"
    >
      Translated&lt;options_dedupe_option_withDupes&gt;
    </label>
  </div>,
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="exactURLMatch"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="exactURLMatch"
    />
    <label
      className="form-check-label"
      htmlFor="exactURLMatch"
    >
      Translated&lt;options_dedupe_option_exactURLMatch&gt;
    </label>
  </div>,
  <div
    className="form-check"
  >
    <input
      checked={true}
      className="form-check-input"
      id="hostnameAndTitleMatch"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="hostnameAndTitleMatch"
    />
    <label
      className="form-check-label"
      htmlFor="hostnameAndTitleMatch"
    >
      Translated&lt;options_dedupe_option_hostnameAndTitleMatch&gt;
    </label>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-8 form-text text-muted"
      style={
        {
          "marginBottom": 0,
        }
      }
    >
      <dl
        style={
          {
            "marginBottom": 0,
          }
        }
      >
        <dt>
          Translated&lt;options_dedupe_option_withDupes_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_withDupes_description&gt;
        </dd>
        <dt
          style={
            {
              "marginTop": "10px",
            }
          }
        >
          Translated&lt;options_dedupe_option_exactURLMatch_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_exactURLMatch_description&gt;
        </dd>
        <dt
          style={
            {
              "marginTop": "10px",
            }
          }
        >
          Translated&lt;options_dedupe_option_hostnameAndTitleMatch_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_hostnameAndTitleMatch_description&gt;
        </dd>
      </dl>
    </div>
  </div>,
]
`;

exports[`TabWrangleOption renders options with withDupes selected 1`] = `
[
  <strong>
    Translated&lt;options_dedupe_label&gt;
  </strong>,
  <div
    className="form-check"
  >
    <input
      checked={true}
      className="form-check-input"
      id="withDupes"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="withDupes"
    />
    <label
      className="form-check-label"
      htmlFor="withDupes"
    >
      Translated&lt;options_dedupe_option_withDupes&gt;
    </label>
  </div>,
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="exactURLMatch"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="exactURLMatch"
    />
    <label
      className="form-check-label"
      htmlFor="exactURLMatch"
    >
      Translated&lt;options_dedupe_option_exactURLMatch&gt;
    </label>
  </div>,
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="hostnameAndTitleMatch"
      name="wrangleOption"
      onChange={[MockFunction]}
      type="radio"
      value="hostnameAndTitleMatch"
    />
    <label
      className="form-check-label"
      htmlFor="hostnameAndTitleMatch"
    >
      Translated&lt;options_dedupe_option_hostnameAndTitleMatch&gt;
    </label>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-8 form-text text-muted"
      style={
        {
          "marginBottom": 0,
        }
      }
    >
      <dl
        style={
          {
            "marginBottom": 0,
          }
        }
      >
        <dt>
          Translated&lt;options_dedupe_option_withDupes_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_withDupes_description&gt;
        </dd>
        <dt
          style={
            {
              "marginTop": "10px",
            }
          }
        >
          Translated&lt;options_dedupe_option_exactURLMatch_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_exactURLMatch_description&gt;
        </dd>
        <dt
          style={
            {
              "marginTop": "10px",
            }
          }
        >
          Translated&lt;options_dedupe_option_hostnameAndTitleMatch_label&gt;
        </dt>
        <dd>
          Translated&lt;options_dedupe_option_hostnameAndTitleMatch_description&gt;
        </dd>
      </dl>
    </div>
  </div>,
]
`;
